test/unit/kubernetes-deploy/kubernetes_resource/deployment_test.rb

```diff
@@ -2,118 +2,332 @@
+      assert deploy.deploy_timed_out?
+    end
   end
 
-  def test_deploy_succeeded_raises_with_invalid_mix_of_annotation
-    rollout = {
-      'spec' => {
-        'strategy' => 'recreate'
-      },
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
+  def test_deploy_timed_out_based_on_progress_deadline_accommodates_stale_conditions_bug_in_k8s_176_and_lower
+    Timecop.freeze do
+      deployment_status = {
+        "replicas" => 3,
+        "conditions" => [{
+          "type" => "Progressing",
+          "status" => 'False',
+          "lastUpdateTime" => Time.now.utc - 5.seconds,
+          "reason" => "Failed to progress"
+        }]
       }
-    }
+      deploy = build_synced_deployment(
+        template: build_deployment_template(status: deployment_status),
+        replica_sets: [build_rs_template(status: { "replica" => 1 })]
+      )
+      deploy.deploy_started_at = Time.now.utc - 5.seconds # progress deadline of 10s has not elapsed
+      deploy.kubectl.expects(:server_version).returns(Gem::Version.new("1.7.6"))
+
+      refute deploy.deploy_timed_out?
+    end
+  end
 
-    kubectl_mock = Minitest::Mock.new
-    status_mock = Minitest::Mock.new
-    status_mock.expect :success?, true
-    kubectl_mock.expect(:run, [true, true, status_mock], [Object, Object, Object, Object, Object, Object])
-    deploy = KubernetesDeploy::Deployment.new(namespace: "", context: "", logger: logger, definition: rollout)
-    deploy.instance_variable_set(:@kubectl, kubectl_mock)
+  private
 
-    refute deploy.validate_definition
+  def build_deployment_template(status: { 'replicas' => 3 }, rollout: nil,
+    strategy: 'rollingUpdate', max_unavailable: nil)
+
+    base_deployment_manifest = fixtures.find { |fixture| fixture["kind"] == "Deployment" }
+    result = base_deployment_manifest.deep_merge("status" => status)
+    result["metadata"]["annotations"][KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION] = rollout if rollout
+
+    if spec_override = status["replicas"].presence # ignores possibility of surge; need a spec_replicas arg for that
+      result["spec"]["replicas"] = spec_override
+    end
+
+    if strategy == "Recreate"
+      result["spec"]["strategy"] = { "type" => strategy }
+    end
+
+    if max_unavailable
+      result["spec"]["strategy"]["rollingUpdate"] = { "maxUnavailable" => max_unavailable }
+    end
+
+    result
+  end
+
+  def build_rs_template(status: { 'replicas' => 3 })
+    base_rs_manifest = fixtures.find { |fixture| fixture["kind"] == "ReplicaSet" }
+    result = base_rs_manifest.deep_merge("status" => status)
+
+    if spec_override = status["replicas"].presence # ignores possibility of surge; need a spec_replicas arg for that
+      result["spec"]["replicas"] = spec_override
+    end
+    result
+  end
+
+  def build_synced_deployment(template:, replica_sets:)
```

Comment:
I would rename this, perhaps:

> build_deployment_from_test_fixture(...)

As a lay person it took me several read throughs to understand that this function creates deployment objects from static test assets.

I think the name indicates that it is syncing something with the cluster.

Reply:
Hm, it doesn't actually build it from a fixture though, now that `build_deployment_template` is a separate helper... It is creating a resource instance, and making that instance look as though `sync` returned the `template` arg. So it is in fact supposed to be simulating a `sync` with the cluster.

Reply:
> So it is in fact supposed to be simulating a sync with the cluster.

Right, it just wasn't immediately obvious which way stuff was syncing (in this case fixture -> object, as opposed to cluster -> object). Feel free to ignore this comment.